### PR TITLE
[Impeller] Add support to embedder for Impeller on GL (via Angle on Windows).

### PR DESCRIPTION
**Pull request by jonahwilliams**
Actually works now, though there is some issue with the default fbo stencil so I've filled https://github.com/flutter/flutter/issues/130048

Other issues:
* ~~Rendering looks wrong~~
* ~~Resizing window hangs~~
* ~~Reactor isn't set up correctly and all blit passes are currently failing.~~
* ~~Needs to handle falling back to sample count of 1 like we do on Android~~.

### CODE REVIEW

File: shell/platform/windows/angle_surface_manager.cc
Comment (jonahwilliams) [inline shell/platform/windows/angle_surface_manager.cc:73]: We could configure this per backend but turning on MSAA for Skia might not be a bad thing to do? w/e TOOD

File: impeller/renderer/backend/gles/render_pass_gles.cc
Comment (jonahwilliams) [inline impeller/renderer/backend/gles/render_pass_gles.cc:453]: This was causing the problem, but haven't actually debugged it yet.

File: shell/platform/embedder/embedder_surface_gl_impeller.cc
Comment (jonahwilliams) [inline shell/platform/embedder/embedder_surface_gl_impeller.cc:183]: This seems wrong but I'm not sure how else to make it work. 🤷‍♂️

File: shell/gpu/gpu_surface_gl_impeller.cc
Comment (jonahwilliams) [inline shell/gpu/gpu_surface_gl_impeller.cc:85]: Adding this is enough to get window resizing working.

File: shell/platform/windows/flutter_windows_engine.cc
Comment (loic-sharma) [inline shell/platform/windows/flutter_windows_engine.cc:381]:
```suggestion
    // Impeller does not support a Software backend. Avoid falling back and
```

File: shell/platform/windows/flutter_windows_engine.cc
Comment (loic-sharma) [inline shell/platform/windows/flutter_windows_engine.cc:383]: Should we `return false` here if surface creation failed and the `surface_manager_` is null?

Comment (loic-sharma) [reply shell/platform/windows/flutter_windows_engine.cc:383]: If yes, I'd also test this behavior here using something similar to this: https://github.com/flutter/engine/blob/1de141c6a6e744b30a1de179370aa0805e4e7b94/shell/platform/windows/flutter_windows_engine_unittests.cc#L173

File: shell/platform/embedder/embedder_surface_gl_impeller.cc
Comment (loic-sharma) [inline shell/platform/embedder/embedder_surface_gl_impeller.cc:89]: Is `ERROR` correct here? I'd expect this to be an `INFO` or `WARNING`

Comment (jonahwilliams) [reply shell/platform/embedder/embedder_surface_gl_impeller.cc:89]: ERROR is the only message that shows up across the board and is what we're using for other platforms to disambiguate whether or not impeller is running. We may remove this at some point in the future.

Comment (loic-sharma) [reply shell/platform/embedder/embedder_surface_gl_impeller.cc:89]: Gotcha, marking as resolved

**Comment by loic-sharma** [comment]
For other folks reviewing, you can test this locally by:

1. Update the tool to support Impeller:

```patch
diff --git a/packages/flutter_tools/lib/src/windows/windows_device.dart b/packages/flutter_tools/lib/src/windows/windows_device.dart
index d94b8bbaef..5fc49d423a 100644
--- a/packages/flutter_tools/lib/src/windows/windows_device.dart
+++ b/packages/flutter_tools/lib/src/windows/windows_device.dart
@@ -40,6 +40,9 @@ class WindowsDevice extends DesktopDevice {
   @override
   String get name => 'Windows';

+  @override
+  bool get supportsImpeller => true;
+
   @override
   Future<TargetPlatform> get targetPlatform async => TargetPlatform.windows_x64;
```

2. Enable impeller in run: `flutter run -d windows --local-engine host_debug_unopt --enable-impeller`

**Comment by yaakovschectman** [comment]
Are we testing this in any way?

**Comment by jonahwilliams** [comment]
we can turn on a devicelab test that runs on impeller, however we've generally kept them in staging (for Android) as we don't want regressions to experimental software to close the tree.

That said, we'll need a devicelab machine that can acquire a hardware surface.

**Comment by jonahwilliams** [comment]
I need to make sure impeller is getting enabled in the test as I expect, but that is waiting for me to get home and get access to a windows machine.

**Comment by jonahwilliams** [comment]
Technically this also allows any other embedder requesting the GL backend to use Impeller too.

File: shell/platform/windows/flutter_windows_engine.cc
Comment (jonahwilliams) [reply shell/platform/windows/flutter_windows_engine.cc]: Done

File: shell/platform/windows/flutter_windows_engine.cc
Comment (jonahwilliams) [reply shell/platform/windows/flutter_windows_engine.cc:388]: Done

File: shell/platform/windows/flutter_windows_engine.cc
Comment (loic-sharma) [inline shell/platform/windows/flutter_windows_engine.cc:387]: This coment is a bit late as at this point we've already checked that we won't use the software backend. Please move this comment up a bit

File: shell/platform/windows/flutter_windows_engine_unittests.cc
Comment (loic-sharma) [inline shell/platform/windows/flutter_windows_engine_unittests.cc:248]: Hm, I believe this test runs in parallel with the `FlutterProjectBundle` tests. For example:

https://github.com/flutter/engine/blob/1b44d7ee1a609cb9743940431704f5f1a50653e5/shell/platform/windows/flutter_project_bundle_unittests.cc#L69-L78

If so, we may run into concurrency issues. It looks like `FlutterProjectBundle` supported overriding the switches using `SetSwitches`, but that's currently broken:

https://github.com/flutter/engine/blob/3f50b7d51678a8899ee390ea22c0ba7a03bcdbef/shell/platform/windows/flutter_project_bundle.cc#L85-L92

Would you mind fixing that `SetSwitches` method and then moving this test to `SetSwitches`? Apologies for the mess. Feel free to skip this if the concurrency concern is invalid. Or also feel free to slap a TODO and I can fix it up in a follow-up change.

**Comment by loic-sharma** [comment]
> I need to make sure impeller is getting enabled in the test as I expect, but that is waiting for me to get home and get access to a windows machine.

Impeller is indeed enabled 😄

File: shell/platform/windows/flutter_windows_engine_unittests.cc
Comment (jonahwilliams) [reply shell/platform/windows/flutter_windows_engine_unittests.cc:248]: Yeah I noticed that but I wasn't sure if this was the correct way to do it or not. If that is supposed to work then happy to fix it, seems like a better way to configure test state..

**Review by loic-sharma — APPROVED** [review]
Windows embedder changes LGTM, but consider getting another approval from someone that's more familiar with the engine surface changes

**Comment by jonahwilliams** [comment]
PTAL @chinmaygarde @dnfield

File: impeller/renderer/backend/gles/render_pass_gles.cc
Comment (dnfield) [inline impeller/renderer/backend/gles/render_pass_gles.cc:466]:
```suggestion
// on Windows causes Angle to discard the entire render target. Until we know
```

File: shell/platform/embedder/embedder.cc
Comment (dnfield) [inline shell/platform/embedder/embedder.cc:76]: Why are these nognchecked?

File: shell/platform/embedder/embedder_surface_gl_impeller.cc
Comment (dnfield) [inline shell/platform/embedder/embedder_surface_gl_impeller.cc:52]: nit: should this be further down after the early returns?

File: shell/platform/embedder/embedder_surface_gl_impeller.cc
Comment (dnfield) [inline shell/platform/embedder/embedder_surface_gl_impeller.cc:83]: e.g. right before this line?

File: shell/platform/embedder/embedder_surface_gl_impeller.cc
Comment (dnfield) [reply shell/platform/embedder/embedder_surface_gl_impeller.cc:180]: I don't think you need this.

You should just see if you can make the resource context current, and if you can mark the worker as allowed to react on this thread. And then always return nullptr.

Returning a valid context from here should not be required and we should not be creating a GrContext when using Impeller.

**Review by dnfield — CHANGES_REQUESTED** [review]
Mostly LG, some nits.

We should do something about the resource context bit before landing this.

File: shell/platform/embedder/embedder_surface_gl_impeller.cc
Comment (jonahwilliams) [reply shell/platform/embedder/embedder_surface_gl_impeller.cc]: Done

File: shell/platform/embedder/embedder_surface_gl_impeller.cc
Comment (jonahwilliams) [reply shell/platform/embedder/embedder_surface_gl_impeller.cc:82]: Done

File: shell/platform/embedder/embedder_surface_gl_impeller.cc
Comment (jonahwilliams) [reply shell/platform/embedder/embedder_surface_gl_impeller.cc]: Ahh OK, that works and makes much more sense. I briefly read through the code and thought this still needed to be constructed

File: impeller/renderer/backend/gles/render_pass_gles.cc
Comment (jonahwilliams) [reply impeller/renderer/backend/gles/render_pass_gles.cc]: Done

File: shell/platform/embedder/embedder.cc
Comment (jonahwilliams) [reply shell/platform/embedder/embedder.cc]: Pushing up a change to remove these to see what complains

**Review by dnfield — APPROVED** [review]
LGTM once CI is happy

File: shell/platform/embedder/embedder_surface_gl_impeller.cc
Comment (dnfield) [reply shell/platform/embedder/embedder_surface_gl_impeller.cc]: The resource context is a performance improving optional thing. Impeller handles things differently by just using worker threads and the reactor mechanism - Skia OTOH expects you to manage a context for the IO thread.

Returning nullptr here should be tolerated anywhere. We may eventually want some better way to signal what this method is actually doing but right now it's a little hard to think about the best way to clean up all the duplicated branches where we've baked Skia concepts into things (like methods that return nullptr for GrContexts in various spots, or nullptr for impeller::Contexts in others).

**Comment by jonahwilliams** [comment]
Ahh the problem was the `#ifdef` instead of `#if`

**Comment by jonahwilliams** [comment]
ahh I guess not, linux fuchsia seems to be trying to enable impeller :(

**Review by chinmaygarde — APPROVED** [review]

**Comment by jonahwilliams** [comment]
@dnfield I put the nognchecks back. I'm not sure exactly how to untangle this, but it seems like its mostly a problem with however we're configuring fuchsia builds. Impeller ends up in a partially enabled state and that is confusing the build rules.

**Pull request merged**
